### PR TITLE
Buttons config v2

### PR DESCRIPTION
**Pull request by hsmett**
image

image

### CODE REVIEW

File: creme/settings.py [hunk 1]
```diff
@@ -8,7 +8,6 @@
 from django.utils.translation import gettext_lazy as _
 
 DEBUG = False
-JAVASCRIPT_DEBUG = DEBUG
```
Comment (genglert) [inline creme/settings.py:11]: I'd prefer a separated patch with an entry in the Changelog ("unused setting XXX was removed ..."

File: creme/creme_config/templatetags/creme_config_tags.py [hunk 2]
```diff
@@ -72,3 +72,8 @@ def config_brick_menu_hide_deleted_cfields_action(context, url, hidden):
         out_label=_('Show deleted custom fields'),
         __value='false' if hidden else 'true',
     )
+
+
+@register.inclusion_tag('creme_config/templatetags/buttons_placeholders.html', takes_context=False)
+def render_buttons_placeholders(buttons, empty_label):
```
Comment (genglert) [inline creme/creme_config/templatetags/creme_config_tags.py:78]: Prefix by 'config_'

File: creme/creme_config/forms/widgets.py [hunk 3]
```diff
@@ -64,3 +64,49 @@ def get_context(self, name, value, attrs):
                 context = button_list.get_context(name=name, value=value, attrs=attrs)
 
         return context
+
+
+class ButtonMenuEditWidget(Widget):
+    template_name = 'creme_config/forms/widgets/buttonmenu-edit-widget.html'
+
```
Comment (genglert) [inline creme/creme_config/forms/widgets.py:71]: 1. I try to remove the old "MyEntityEditForm" classes to replace them by "MyEntityEditionForm". So for consistency 'ButtonMenuEditionWidget' would be better.
2. Hum in 'creme_core/forms/widgets' we found XXCreatorWidget, XXXSelector, XXXEditionWidget. That's not very consistent, but the last pattern matches my remark in 1. And so ButtonMenuEditor & ButtonMenuEditorWidget could be used to. Choose one.
3. Don't forget to rename the template too ; even if you keep the 'Widget' suffix in the class, remove it from the file-name (like TimeWidget => 'widgets/time.html')

File: creme/creme_config/forms/widgets.py [hunk 4]
```diff
@@ -64,3 +64,49 @@ def get_context(self, name, value, attrs):
...
+    def __init__(self, attrs=None, choices=()):
+        super().__init__(attrs)
+        # choices can be any iterable, but we may need to render this widget
+        # multiple times. Thus, collapse it into a list so it can be consumed
+        # more than once.
+        self.choices = list(choices)
```
Comment (genglert) [inline creme/creme_config/forms/widgets.py:77]: self.choices = [*choices]

File: creme/creme_config/templates/creme_config/bricks/button-menu.html [hunk 5]
```diff
@@ -11,7 +11,7 @@
 {% has_perm_to admin 'creme_core' as admin_perm %}{% url 'creme_config__add_buttons_to_ctype' as add_url %}
     {% brick_header_action id='add' url=add_url label=_('New buttons configuration') enabled=admin_perm %}
 {% endblock %}
-
+{% load creme_config_tags %}
```
Comment (genglert) [inline creme/creme_config/templates/creme_config/bricks/button-menu.html:14]: At the beginning of the file, with other loads.

File: creme/creme_config/bricks.py [hunk 6]
```diff
@@ -634,25 +635,35 @@ class ButtonMenuBrick(Brick):
     template_name = 'creme_config/bricks/button-menu.html'
     configurable = False
 
-    def detailview_display(self, context):
+    @staticmethod
+    def get_buttons():
+        get_ct = ContentType.objects.get_for_id
```
Comment (genglert) [inline creme/creme_config/bricks.py:640]: The field 'ButtonMenuItem.content_type' is a CTypeForeignKey, which automatically use ContentType.objects.get_for_id(). So this optimzation is useless.

**Review by genglert — CHANGES_REQUESTED** [review]

File: creme/creme_config/forms/button_menu.py [hunk 7]
```diff
@@ -24,12 +24,14 @@
...
-from creme.creme_core.forms.widgets import OrderedMultipleChoiceWidget
 from creme.creme_core.gui import button_menu
 from creme.creme_core.models import ButtonMenuItem
-# from creme.creme_core.utils.id_generator import generate_string_id_and_save
 from creme.creme_core.utils.unicode_collation import collator
 
+from .widgets import ButtonMenuEditionWidget
+
+# from creme.creme_core.utils.id_generator import generate_string_id_and_save
+
```
[the remark this answers is not in the document]
Comment (hsmett) [reply creme/creme_config/forms/button_menu.py:34]: I am surprised too, I don't know how it happened. I'm moving it.

File: creme/creme_config/tests/test_widgets.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+from collections import namedtuple
+
+from creme.creme_config.forms.widgets import ButtonMenuEditionWidget
+from creme.creme_core.tests.base import CremeTestCase
+
+Option = namedtuple("Option", ['id_', 'verbose_name', 'description'])
+
+
+class ButtonMenuEditWidgetTestCase(CremeTestCase):
```
Comment (genglert) [inline creme/creme_config/tests/test_widgets.py:11]: ButtonMenuEdit-ION-WidgetTestCase to be consistent

**Review by genglert — CHANGES_REQUESTED** [review]

File: creme/creme_config/locale/fr/LC_MESSAGES/django.po [hunk 9]
```diff
@@ -354,6 +354,16 @@ msgstr "Lignes"
 msgid "Buttons to display"
 msgstr "Boutons à afficher"
 
+#: forms/button_menu.py:110
+msgid ""
+"Drag and drop the buttons between the available buttons and the selected "
+"buttons sections to enable or disable the buttons, or within the selected "
+"buttons section to change the order."
+msgstr "Faites glisser et déposez les boutons entre la section des boutons "
+"disponible et celle des boutons sélectionnés pour activer ou désactiver "
```
Comment (genglert) [inline creme/creme_config/locale/fr/LC_MESSAGES/django.po:363]: disponibleS

**Review by genglert — CHANGES_REQUESTED** [review]

File: creme/creme_config/static/creme_config/js/widgets.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+(function($) {
+    "use strict";
+
+    function buildButtonChoice(choice) {
+        return $("<span>").addClass("menu_button").text(choice.label).attr('title', choice.description).append(
+            $("<input>").attr({
+                type: "hidden",
+                name: choice.name
+            }).val(choice.value).prop("disabled", !choice.selected)
+        );
```
Comment (joehybird) [inline creme/creme_config/static/creme_config/js/widgets.js:11]: There is a String.prototype.template method in creme similar to an extremely simplified es6 one.
```suggestion
return $((
    '<span class="menu_button" title="${description}">' +
        '${label}<input type="hidden" name="${name} val="${value}"/>' +
    '</span>'
).template(choice)).prop("disabled", !choice.selected);
```

File: creme/creme_config/static/creme_config/js/widgets.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    function buildWidget(widget, choices) {
+        choices.map(
```
Comment (joehybird) [inline creme/creme_config/static/creme_config/js/widgets.js:15]: `.each(` will do the job no ? :smile:

File: creme/creme_config/templates/creme_config/forms/widgets/buttonmenu-editor.html [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+  </div>
+  <div class="widget-selected buttons-list instance-buttons">
+    <div class="buttons-list-header">{% trans "Selected buttons" %}</div>
+    <div class="widget-container"></div>
+  </div>
+</div>
+
+{{ widget.choices|json_script:"buttons-widget-choices" }}
+
+<script type="text/javascript">
```
Comment (joehybird) [inline creme/creme_config/templates/creme_config/forms/widgets/buttonmenu-editor.html:16]: Hum... you can create a creme widget instead off initialize it in scripts. But if is the widget is used once, why not.

File: creme/creme_config/templates/creme_config/forms/widgets/buttonmenu-editor.html [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    $(document).ready(function() {
+        creme.initButtonMenuWidget({
+            optionsId: 'buttons-widget-choices',
+            widgetAttrsId: '{{widget.attrs.id}}'
```
Comment (joehybird) [inline creme/creme_config/templates/creme_config/forms/widgets/buttonmenu-editor.html:20]: Please pass directly the jquery element to the "constructor"
```
creme.initButtonMenuWidget($('#{{widget.attrs.id}}'), { ... })
```

File: creme/creme_config/static/creme_config/js/widgets.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    creme.initButtonMenuWidget = function (options) {
+        var buttonsWidgetOptgroups = JSON.parse(document.getElementById(options.optionsId).textContent);
+
+        var widget = $("#" + options.widgetAttrsId);
+
+        buildWidget(widget, buttonsWidgetOptgroups);
+
+        new Sortable(  // eslint-disable-line
```
Comment (joehybird) [inline creme/creme_config/static/creme_config/js/widgets.js:37]: I guess that the 'new not returned' error was annoying you :smile:

File: creme/creme_config/static/creme_config/js/widgets.js [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            widget.find(".widget-available .widget-container").get(0),
+            {
+            group: options.widgetAttrsId,
+            animation: 150,
+            sort: false,
+            onSort: function eventHandler(event) {
```
Comment (joehybird) [inline creme/creme_config/static/creme_config/js/widgets.js:43]: A bit repetitive, can you factorize this function ?

**Review by joehybird — CHANGES_REQUESTED** [review]

**Review by genglert — APPROVED** [review]

**Review by joehybird — APPROVED** [review]

**Pull request merged**